rbd: add generic ephemeral volume validation

humblec opened the pull request:
rbd: add generic ephemeral volume validation
    
    This commit adds the validation of csi RBD driver to work with
    ephemeral volume support. With ephemeral volume support a user can
    specify ephemeral volumes in its pod spec and tie the lifecycle
    of the PVC with the POD.
    
    An example POD spec looks like this:
    
    kind: Pod
    apiVersion: v1
    metadata:
      name: my-app
       spec:
      containers:
        - name: csi-rbd-demo-pod
          image: busybox
          volumeMounts:
            - mountPath: "/scratch"
              name: mypvc
          command: [ "sleep", "1000000" ]
      volumes:
        - name: mypvc
          ephemeral:
            volumeClaimTemplate:
              spec:
                accessModes: [ "ReadWriteOnce" ]
                storageClassName: "csi-rbd-sc"
                resources:
                  requests:
                    storage: 1Gi

Additional Ref # https://github.com/rook/rook/pull/9055

Updates #2587 

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -367,6 +368,28 @@ var _ = Describe("RBD", func() {
+				// create application
+				app, err := loadApp(appEphemeralPath)
+				if err != nil {
+					e2elog.Failf("failed to load application with error %v", err)
```
remove with error from all places

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -367,6 +368,28 @@ var _ = Describe("RBD", func() {
+					e2elog.Failf("failed to delete application with error %v", err)
+				}
+				// validate created backend rbd images
+				validateRBDImageCount(f, 0, defaultRBDPool)
```
please validate image in trash

Madhu-1 commented on examples/rbd/pod-ephemeral.yaml:
```diff
@@ -0,0 +1,22 @@
+        volumeClaimTemplate:
+          spec:
+            accessModes: ["ReadWriteOnce"]
+            storageClassName: "csi-rbd-sc"
```
quotes not required for mountPath and storageclass and also for access mode? please keep the same as in other yamls.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -367,6 +368,28 @@ var _ = Describe("RBD", func() {
 					}
 				})
 			}
+			By("verify ephemeral volume support", func() {
```
generic ephemeral volume support?

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -367,6 +368,28 @@ var _ = Describe("RBD", func() {
 			}
+			By("verify ephemeral volume support", func() {
+				// create application
+				app, err := loadApp(appEphemeralPath)
```
CI will fail for 20 as the feature gate is still alpha?

* add kube validation check here
* enable feature gate for kube 1.20?

humblec replied:
running the tests only >= 21 is good enough I believe. wdyt?

Madhu-1 replied:
my suggestion is to run from 20+ as we test on kube 1.20,21, and 22 also

humblec replied:
on more thoughts, considering other tests are enabled for beta for example, snapshot/clone...etc it looks like its fine to enable only on versions 1.21. also next release is anyway going to throw away 1.20 from here, so should be good I believe. Last patch refresh adjusted the same.

Madhu-1 commented:
please add same e2e for cephfs also

humblec committed:
rbd: add generic ephemeral volume validation

This commit adds the validation of csi RBD driver to work with
ephemeral volume support. With ephemeral volume support a user can
specify ephemeral volumes in its pod spec and tie the lifecycle
of the PVC with the POD.

An example pod spec is also included in this commit.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
cephfs: add validation for generic ephemeral volumes

This commit adds the validation of csi cephfs driver to work with
ephemeral volume support. With ephemeral volume support a user can
specify ephemeral volumes in its pod spec and tie the lifecycle
of the PVC with the POD.

An example POD spec also included in this commit.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
@Madhu-1 added tests for cephfs too and enabled feature gate only for kube versions >=1.20.  PTAL.

nixpanic commented on scripts/minikube.sh:
```diff
@@ -202,6 +211,14 @@ up)
+    #  get kubernetes version we are operating on and accordingly enable feature gates
+    KUBE_MAJOR=$(kube_version 1)
+    KUBE_MINOR=$(kube_version 2)
+    if  [ "${KUBE_MAJOR}" -eq 1 ] && [ "${KUBE_MINOR}" -ge 20 ];
```
1. not sure you can compare strings with `-eq` or `-ge`
2. no `;` needed when `then` is on the next  line

humblec replied:
the `-eq or ge` is working as expected, I have removed unwanted `;` in the last refresh.

nixpanic commented on scripts/minikube.sh:
```diff
@@ -202,6 +211,14 @@ up)
+    if  [ "${KUBE_MAJOR}" -eq 1 ] && [ "${KUBE_MINOR}" -ge 20 ];
+    then
+    # if kubernetes version is greater than 1.20 enable Generic Ephemeral Volume feature gate
+    K8S_FEATURE_GATES="GenericEphemeralVolume=true"
```
`K8S_FEATURE_GATES` is already set at line 163, this overrides that option. Should the variable not be extended with the feature gate instead?

humblec replied:
my plan was to remove the expandCSIVolumes feature gate as a follow up as it no longer wanted, will file a seperate PR for the same, however its better to be in appended string form here, so that in future if there is already set value, it should carry that in final string.

nixpanic reviewed (changes requested):
It is simpler to included the example .yaml files with the e2e commits so that they are self contained and really belong together. There is then no need to include the full yaml in the commit message anymore, cleaning things up a little.

humblec commented:
> It is simpler to included the example .yaml files with the e2e commits so that they are self contained and really belong together. There is then no need to include the full yaml in the commit message anymore, cleaning things up a little.

sure, squashed commits of example yaml to the validation commit.

nixpanic commented on scripts/minikube.sh:
```diff
@@ -202,6 +211,14 @@ up)
+    if  [ "${KUBE_MAJOR}" -eq 1 ] && [ "${KUBE_MINOR}" -ge 20 ]
+    then
+    # if kubernetes version is greater than 1.20 enable Generic Ephemeral Volume feature gate
+    K8S_FEATURE_GATES="${K8S_FEATURE_GATES},GenericEphemeralVolume=true"
```
if `K8S_FEATURE_GATES` is an empty string, maybe the extra `,` can cause issues?

humblec replied:
as discussed, its not causing issue rather its going through, so will keep this version.

nixpanic commented on scripts/minikube.sh:
```diff
@@ -160,7 +169,7 @@ CSI_RESIZER_VERSION=${CSI_RESIZER_VERSION:-"v1.2.0"}
 
 #feature-gates for kube
-K8S_FEATURE_GATES=${K8S_FEATURE_GATES:-"ExpandCSIVolumes=true"}
+K8S_FEATURE_GATES=${K8S_FEATURE_GATES:-""}
```
Now you are dropping `ExpandCSIVolumes` without comment. It is better to do that in a stand-alone commit. As a cleanup, it can be included in this PR, no need to create a separate one.

humblec replied:
sure, put it as a seperate commit here in this PR itself.

humblec committed:
deploy: remove expandCSIVolumes feature gate from deployment

The `expandCSIVolumes` feature gate is beta since kubernetes 1.16
version and we no longer wanted to explictly enable it in the
deployment.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.20

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest  ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
@Mergifyio rebase

humblec commented:
/retest all

humblec commented:
/retest all

humblec commented:
@Mergifyio refresh

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.20

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
Test failure https://github.com/ceph/ceph-csi/issues/2610

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.20

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
@Madhu-1 can you please review this ?

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -312,6 +313,36 @@ var _ = Describe("cephfs", func() {
+				}
+				err := createCephfsStorageClass(f.ClientSet, f, true, nil)
+				if err != nil {
+					e2elog.Failf("failed to create CephFS storageclass with error %v", err)
```
`failed to create CephFS storageclass with error %v` to `failed to create CephFS storageclass: %v` `with error` formatting needs to be adjusted in other places also

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -312,6 +313,36 @@ var _ = Describe("cephfs", func() {
+				err = createApp(f.ClientSet, app, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to create application: %v", err)
+				}
```
add a subvolumeCount validation here to make sure subvolume is present in backend?

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -367,6 +368,40 @@ var _ = Describe("RBD", func() {
+				}
+				// validate created backend rbd images
+				validateRBDImageCount(f, 0, defaultRBDPool)
+				// validate images in trash
+				err = waitToRemoveImagesFromTrash(f, defaultRBDPool, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to validate rbd images in pool %s trash: %v", defaultRBDPool, err)
+				}
+				err = waitToRemoveImagesFromTrash(f, defaultRBDPool, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to validate rbd images in pool %s trash: %v", defaultRBDPool, err)
+				}
+			})
+
```
```suggestion
// validate images in trash
				err = waitToRemoveImagesFromTrash(f, defaultRBDPool, deployTimeout)
				if err != nil {
					e2elog.Failf("failed to validate rbd images in pool %s trash: %v", defaultRBDPool, err)
				}
```

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -202,6 +211,14 @@ up)
+    if  [ "${KUBE_MAJOR}" -eq 1 ] && [ "${KUBE_MINOR}" -ge 21 ]
+    then
+    # if kubernetes version is greater than 1.21 enable Generic Ephemeral Volume feature gate
+    K8S_FEATURE_GATES="${K8S_FEATURE_GATES},GenericEphemeralVolume=true"
```
as it is beta in 1.21+ already enabled by default. we don't need to enable it explicitly?

humblec replied:
@Madhu-1 if this script is executed for versions less than 1.21, we dont need to enable this feature gate, thats why tihs check in place.

humblec replied:
it was added considering the script can be used independently for kube deployments, but may be here, we can skip that and only have e2e check. adjusted the PR..

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -312,6 +313,36 @@ var _ = Describe("cephfs", func() {
 			}
+			By("verify generic ephemeral volume support", func() {
+				// generic ephemeral volume support is beta since v1.21.
+				if !k8sVersionGreaterEquals(f.ClientSet, 1, 21) {
```
it's skipping for 1.20 but as per https://github.com/ceph/ceph-csi/pull/2603#discussion_r739026205 it's agreed to run on  1.20 also?

humblec replied:
this was deliberate, considering this is beta and supported since 1.21, this should be good I believe.

nixpanic replied:
agreed, but in that case it is not needed to enable the featue gate in `minikube.sh`?

humblec replied:
it was added considering the script can be used independently for kube deployments, but may be here, we can skip that and only have e2e check. adjusted the PR..

humblec commented:
@nixpanic @Madhu-1 all the comments are addressed, ptal.. thanks in advance.

nixpanic commented on examples/rbd/pod-ephemeral.yaml:
```diff
@@ -0,0 +1,22 @@
+      ephemeral:
+        volumeClaimTemplate:
+          spec:
+            accessModes: [ReadWriteOnce]
```
The other examples use the cleaner (IMHO) format of

```yaml
accessModes:
  - ReadWriteOnce
```

it is probably better to use the same formatting in all examples we have

nixpanic reviewed (dismissed):
just a nit about the yaml formatting of the examples, looks good to me otherwise

Madhu-1 commented on examples/rbd/pod-ephemeral.yaml:
```diff
@@ -0,0 +1,23 @@
+  name: csi-rbd-demo-ephemeral-pod
+spec:
+  containers:
+      image: "docker.io/library/nginx:latest"
```
remove `""` from image name

Madhu-1 commented on examples/rbd/pod-ephemeral.yaml:
```diff
@@ -0,0 +1,23 @@
+              requests:
+                storage: 1Gi
+            storageClassName: csi-rbd-sc
+      name: mypvc
```
move the name to line 14  and keep it similar to cephfs

humblec commented:
@Mergifyio rebase

humblec commented:
@nixpanic @Madhu-1 ptal.. thanks.

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
rebased to devel, and also a minor change ( as pointed out by @Rakshith-R in ci logs)  from `appEphemeralPath = cephFSExamplePath + "pod-ephemeral.yaml"` to `appEphemeralPath := cephFSExamplePath + "pod-ephemeral.yaml"`

humblec commented:
@Mergifyio rebase

nixpanic commented:
/retest ci/centos/upgrade-tests

nixpanic commented:
> /retest ci/centos/upgrade-tests

For some reason the CentOS jobs were not listed, this causes a refresh from Jenkins in the CentOS CI.

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.20

humblec commented:
@Mergifyio rebase

humblec commented:
@Mergifyio rebase

humblec commented:
@Mergifyio refresh

humblec commented:
@Mergifyio rebase

Rakshith-R commented:
@mergifyio rebase

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
@Mergifyio rebase

humblec commented:
@Mergifyio rebase

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21
